Felienne/hedy
# Review user-facing type error messages; Refactor comparisons #1298

**Description**

- Revise all user-facing type error messages
- Refactor shamelessly comparisons' grammar and tests
- Add type validation for comparisons, add to list, remove from list commands, in list and equality checks

**Fix for**

#1298

**How to test**

1. Check whether the new user-facing messages are ok:
- In level 1 run the command `forward ten`. This is the general exception telling the user that they are using a type that is does not work for the command at hand. Is the error message clear?
- In level 2 run the command `a is ten\n forward a`. While this is the same error, it has a slightly different twist because now it is not referring to a literal value (as in the previous example), but to a variable. Is the error message clear?
- In level 12 run `x is 'test1'\n y is x + 1`. This is the error indicating that both types are allows with the command but they are not compatible with each other. Is it clear?
- In level 2 run `turn right 90`. This is a variation of the ParseError: `right 90` is something wrong and it does not really have a type. However, we suggest that it is changed to something that is supported. Is it clear?

2. Check that the grammar change of level 14 is correct. The change is solely  done based on assumptions made from the available tests.

**Checklist**

If you're unsure about any of these, don't hesitate to ask. We're here to help!
